packages/api/amplify_api/lib/src/api_plugin_impl.dart

```diff
@@ -0,0 +1,397 @@
+import 'package:amplify_api/src/util/amplify_authorization_rest_client.dart';
+import 'package:amplify_core/amplify_core.dart';
+import 'package:flutter/services.dart';
+import 'package:meta/meta.dart';
+
+/// {@template amplify_api.amplify_api_dart}
+/// The AWS implementation of the Amplify API category.
+/// {@endtemplate}
+class AmplifyAPIDart extends AmplifyAPI {
+  /// {@macro amplify_api.amplify_api_dart}
+  AmplifyAPIDart({
+    List<APIAuthProvider> authProviders = const [],
+    AWSHttpClient? baseHttpClient,
+    this.modelProvider,
+    this.subscriptionOptions,
+  })  : _baseHttpClient = baseHttpClient,
+        super.protected() {
+    authProviders.forEach(registerAuthProvider);
+  }
+
+  late final AWSApiPluginConfig _apiConfig;
+  final AWSHttpClient? _baseHttpClient;
+  late final AmplifyAuthProviderRepository _authProviderRepo;
+  final _logger = AmplifyLogger.category(Category.api);
+
+  /// A map of the keys from the Amplify API config with auth modes to HTTP clients
+  /// to use for requests to that endpoint/auth mode. e.g. { "myEndpoint.AWS_IAM": AWSHttpClient}
+  final Map<String, AWSHttpClient> _clientPool = {};
+
+  /// A map of the keys from the Amplify API config websocket connections to use
+  /// for that endpoint.
+  final Map<String, WebSocketConnection> _webSocketConnectionPool = {};
+
+  /// The registered [APIAuthProvider] instances.
+  final Map<APIAuthorizationType, APIAuthProvider> _authProviders = {};
+
+  /// Subscription options
+  final GraphQLSubscriptionOptions? subscriptionOptions;
+
+  @override
+  Future<void> configure({
+    AmplifyConfig? config,
+    required AmplifyAuthProviderRepository authProviderRepo,
+  }) async {
+    final apiConfig = config?.api?.awsPlugin;
+    if (apiConfig == null) {
+      throw const ApiException(
+        'No AWS API config found',
+        recoverySuggestion: 'Add API from the Amplify CLI. See '
+            'https://docs.amplify.aws/lib/graphqlapi/getting-started/q/platform/flutter/#configure-api',
+      );
+    }
+    _apiConfig = apiConfig;
+    _authProviderRepo = authProviderRepo;
+    _registerApiPluginAuthProviders();
+  }
+
+  /// Register AmplifyAuthProviders that are specific to API category (API key,
+  /// OIDC or Lambda).
+  ///
+  /// If an endpoint has an API key, ensure valid auth provider registered.
+  ///
+  /// Register OIDC/Lambda set to _authProviders in constructor.
+  void _registerApiPluginAuthProviders() {
+    _apiConfig.endpoints.forEach((key, value) {
+      // Check the presence of apiKey (not auth type) because other modes might
+      // have a key if not the primary auth mode.
+      if (value.apiKey != null) {
+        _authProviderRepo.registerAuthProvider(
+          APIAuthorizationType.apiKey.authProviderToken,
+          AppSyncApiKeyAuthProvider(),
+        );
+      }
+    });
+
+    // Register OIDC/Lambda auth providers.
+    for (final authProvider in _authProviders.values) {
```

Comment:
Behavior is also unclear here when multiple providers are passed. Also, no  distinction between OIDC/Lambda is being made even though they're two different auth types.